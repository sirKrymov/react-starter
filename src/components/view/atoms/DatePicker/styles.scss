/* stylelint-disable selector-max-compound-selectors, max-nesting-depth */
@import '../../../../styles/configs';
@import '../../../../styles/mixins';

.date-picker {
  min-height: inherit;
  position: relative;

  .react-datepicker-wrapper {
    min-height: inherit;
    width: 100%;

    .react-datepicker {
      &__input-container {
        min-height: inherit;

        input {
          -webkit-appearance: none;
          background-color: var(--bgPrimary);
          border-radius: $border-radius;
          transition: border-color $anim-time-fast,
            background-color $anim-time-fast, opacity $anim-time;
          box-shadow: none;
          min-height: inherit;
          padding: 0 16px;
          outline: none;
          border: 1px solid var(--borderPrimary);
          color: var(--textPrimary);
          width: 100%;

          &::placeholder {
            color: var(--textHint);
          }

          &:focus {
            border-color: var(--thirdy);
            box-shadow: none;
          }
        }
      }
    }
  }

  &--right-block-xs {
    .react-datepicker-wrapper {
      .react-datepicker__input-container {
        input {
          padding-right: 28px;
        }
      }
    }
  }

  &--right-block-sm {
    .react-datepicker-wrapper {
      .react-datepicker__input-container {
        input {
          padding-right: 36px;
        }
      }
    }
  }

  &--right-block-md {
    .react-datepicker-wrapper {
      .react-datepicker__input-container {
        input {
          padding-right: 44px;
        }
      }
    }
  }

  &--right-block-lg {
    .react-datepicker-wrapper {
      .react-datepicker__input-container {
        input {
          padding-right: 52px;
        }
      }
    }
  }

  &--error {
    .react-datepicker-wrapper {
      .react-datepicker__input-container {
        input {
          border-color: var(--error);
        }
      }
    }
  }

  &__right-block {
    transform: translateY(-50%);
    position: absolute;
    height: 100%;
    right: 0;
    top: 50%;
    @include center;

    &--xs {
      width: 28px;
    }

    &--sm {
      width: 36px;
    }

    &--md {
      width: 44px;
    }

    &--lg {
      width: 52px;
    }
  }
}
