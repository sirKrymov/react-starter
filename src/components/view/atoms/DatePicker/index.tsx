// @SETUP: Please don't forget to remove this component or current comment if component is used in project
import React, { FocusEvent, ReactElement } from 'react';

import { IoMdCloseCircle, IoIosCalendar } from 'react-icons/io';
import ReactDatePicker from 'react-datepicker';
import moment from 'moment';
import cn from 'classnames';

import { ReactIcon } from 'components/view/atoms/ReactIcon';
import { UiSize } from 'components/view/atoms/UiSize';

import './styles.scss';

export interface IProps {
  getCustomDateFormat?(date: Date): string | any;
  onBlur?(e: FocusEvent<any>): void;
  onChange(date: string): void;
  customConfig?: Record<string, any>;
  placeholder?: string;
  className?: string;
  onClick?(): void;
  onClear?(): void;
  disabled?: boolean;
  height?: 'xs' | 'sm' | 'md' | 'lg' | 'full';
  width?: 'xs' | 'sm' | 'md' | 'lg' | 'xl' | 'full' | 'content';
  error?: boolean;
  value?: string;
  name: string;
  id: string;
}

export function DatePicker({
  getCustomDateFormat,
  customConfig,
  placeholder,
  className,
  disabled,
  onChange,
  onClear,
  onBlur,
  height = 'md',
  width,
  error,
  value,
  name,
  id
}: IProps): ReactElement {
  const config = {
    dateFormat: 'yyyy-MM-dd',
    ...customConfig
  };

  function handleChange(date: Date): void {
    const modifiedDate = date
      ? !!getCustomDateFormat
        ? getCustomDateFormat(date)
        : moment(date).format()
      : '';

    onChange(modifiedDate);
  }

  const datePickerClass = cn(
    {
      'date-picker': true,
      [`date-picker--right-block-${height}`]: height,
      'date-picker--error': error
    },
    className
  );

  const rightBlockClass = cn({
    'input__right-block': true,
    [`input__right-block--${height}`]: height
  });

  return (
    <UiSize width={width} height={height}>
      <div className={datePickerClass} id={id}>
        <ReactDatePicker
          placeholderText={placeholder}
          showPopperArrow={false}
          onChange={handleChange}
          disabled={disabled}
          selected={(value && new Date(value)) || undefined}
          onBlur={onBlur}
          name={name}
          {...config}
        />

        <div className={rightBlockClass}>
          {value && onClear ? (
            <ReactIcon onClick={onClear} size="md">
              <IoMdCloseCircle />
            </ReactIcon>
          ) : (
            <ReactIcon>
              <IoIosCalendar />
            </ReactIcon>
          )}
        </div>
      </div>
    </UiSize>
  );
}
