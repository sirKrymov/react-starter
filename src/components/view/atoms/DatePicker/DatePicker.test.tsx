import React from 'react';

import { IoMdCloseCircle } from 'react-icons/io';

import { shallowToJson } from 'enzyme-to-json';

import renderer, { ReactTestRenderer } from 'react-test-renderer';
import { mount, shallow, ShallowWrapper } from 'enzyme';
import moment from 'moment';

// import ReactDatePicker from 'react-datepicker';

// import sinon from 'sinon';

import { DatePicker } from '../DatePicker';

describe('<Input> atom component', () => {
  const getCustomDateFormatMock = jest.fn();
  const handleChange = jest.fn();
  const handleClear = jest.fn();

  let shallowComponent: ShallowWrapper;
  let mountComponent: any;
  let component: ReactTestRenderer;

  const createComponent = ({
    onChange,
    value,
    ...rest
  }: Record<string, any>) => {
    component = renderer.create(
      <DatePicker
        onChange={onChange}
        value={value}
        name="date-picker"
        id="check-datepicker"
        {...rest}
      />
    );
  };

  const createMountComponent = ({
    onChange,
    value,
    ...rest
  }: Record<string, any>) => {
    mountComponent = mount(
      <DatePicker
        onChange={onChange}
        value={value}
        name="date-picker"
        id="check-datepicker"
        {...rest}
      />
    );
  };

  const createShallowComponent = ({
    onChange,
    value,
    ...rest
  }: Record<string, any>) => {
    shallowComponent = shallow(
      <DatePicker
        onChange={onChange}
        value={value}
        name="date-picker"
        id="check-datepicker"
        {...rest}
      />
    );
  };

  describe('WHEN the component loads', () => {
    beforeEach(() => {
      createComponent({ onChange: handleChange, value: '' });
    });

    it('THEN it should render without crashing', () => {
      expect(component.toJSON()).toBeDefined();
    });

    it('THEN it should match snapshot', () => {
      expect(component.toJSON()).toMatchSnapshot();
    });
  });

  describe('WHEN the component have different props', () => {
    it('THEN it should fires on change correctly', () => {
      createMountComponent({ onChange: handleChange, value: '' });

      const event = { target: { value: moment('2018-01-22') } };

      expect(mountComponent.find('input').simulate('change', event));
      expect(handleChange).toHaveBeenCalledTimes(1);
      expect(handleChange).toHaveBeenCalledWith('2018-01-22T00:00:00+02:00');
    });

    it('THEN it should render date input correctly with date value', () => {
      createMountComponent({ onChange: handleChange, value: '2018-01-22' });

      expect(mountComponent.find('input').prop('value')).toEqual('2018-01-22');
    });

    it('THEN it should render clear icon and logic', () => {
      createShallowComponent({
        onChange: handleChange,
        onClear: handleClear,
        value: '2018-01-22'
      });

      expect(shallowComponent.find(IoMdCloseCircle)).toBeDefined();
    });

    it('THEN it should custom format date value', () => {
      createShallowComponent({
        getCustomDateFormat: getCustomDateFormatMock,
        onChange: handleChange,
        value: '2018-01-22'
      });

      // getCustomDateFormatMock.mockReturnValue('2018/01/22');

      // getCustomDateFormatMock();

      // expect(handleChange).toHaveBeenCalledWith('2018/01/22');
    });

    it('THEN it should match snapshot', () => {
      createShallowComponent({
        getCustomDateFormat: getCustomDateFormatMock,
        onChange: handleChange,
        value: '2018-01-22'
      });

      expect(shallowToJson(shallowComponent)).toMatchSnapshot();
    });
  });
});
