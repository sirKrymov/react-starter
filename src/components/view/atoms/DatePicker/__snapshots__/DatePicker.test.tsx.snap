// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input> atom component WHEN the component have different props THEN it should match snapshot 1`] = `
<UiSize
  height="md"
>
  <div
    className="date-picker date-picker--right-block-md"
    id="check-datepicker"
  >
    <o
      allowSameDay={false}
      customTimeInput={null}
      dateFormat="yyyy-MM-dd"
      dateFormatCalendar="LLLL yyyy"
      disabled={false}
      disabledKeyboardNavigation={false}
      dropdownMode="scroll"
      enableTabLoop={true}
      excludeScrollbar={true}
      inlineFocusSelectedMonth={false}
      monthsShown={1}
      name="date-picker"
      nextMonthButtonLabel="Next Month"
      nextYearButtonLabel="Next Year"
      onBlur={[Function]}
      onCalendarClose={[Function]}
      onCalendarOpen={[Function]}
      onChange={[Function]}
      onClickOutside={[Function]}
      onFocus={[Function]}
      onInputClick={[Function]}
      onInputError={[Function]}
      onKeyDown={[Function]}
      onMonthChange={[Function]}
      onSelect={[Function]}
      onYearChange={[Function]}
      preventOpenOnFocus={false}
      previousMonthButtonLabel="Previous Month"
      previousYearButtonLabel="Previous Year"
      readOnly={false}
      renderDayContents={[Function]}
      selected={2018-01-22T00:00:00.000Z}
      shouldCloseOnSelect={true}
      showFullMonthYearPicker={false}
      showMonthYearPicker={false}
      showPopperArrow={false}
      showPreviousMonths={false}
      showQuarterYearPicker={false}
      showTimeInput={false}
      showTimeSelect={false}
      showYearPicker={false}
      strictParsing={false}
      timeCaption="Time"
      timeInputLabel="Time"
      timeIntervals={30}
      withPortal={false}
    />
    <div
      className="input__right-block input__right-block--md"
    >
      <ReactIcon>
        <IoIosCalendar />
      </ReactIcon>
    </div>
  </div>
</UiSize>
`;

exports[`<Input> atom component WHEN the component loads THEN it should match snapshot 1`] = `
<div
  className="ui-size ui-size--height-md ui-size--width-full"
>
  <div
    className="date-picker date-picker--right-block-md"
    id="check-datepicker"
  >
    <div
      className="react-datepicker-wrapper"
    >
      <div
        className="react-datepicker__input-container"
      >
        <input
          className=""
          disabled={false}
          name="date-picker"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          readOnly={false}
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="input__right-block input__right-block--md"
    >
      <div
        className="react-icon react-icon--color-primary react-icon--size-lg"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M424 96h-40v24c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8V96H160v24c0 4.4-3.6 8-8 8h-16c-4.4 0-8-3.6-8-8V96H88c-22 0-40 18-40 40v272c0 22 18 40 40 40h336c22 0 40-18 40-40V136c0-22-18-40-40-40zm8 300c0 11-9 20-20 20H100c-11 0-20-9-20-20V216c0-4.4 3.6-8 8-8h336c4.4 0 8 3.6 8 8v180zM160 72c0-4.4-3.6-8-8-8h-16c-4.4 0-8 3.6-8 8v24h32V72zM384 72c0-4.4-3.6-8-8-8h-16c-4.4 0-8 3.6-8 8v24h32V72z"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
